File: be-nodejs/src/services/folder.service.ts
import { FolderRequest } from "@/types/req/folder.type";
import { FolderRepository } from "../repositories/folder.repo";
import { Folder } from "@/models/folder.model";
import { PatientService } from "./patient.service";
import { ChatSessionInfo, FolderResponse } from "@/types/res/folder.res";
import { ChatSessionRepository } from "@/repositories/chatsession.repo";

export class FolderService {


    private folderRepository: FolderRepository;
    private patientService: PatientService;
    private chatSessionRepository: ChatSessionRepository;

    constructor() {
        this.patientService = new PatientService();
        this.folderRepository = new FolderRepository();
        this.chatSessionRepository = new ChatSessionRepository();
    }

    
    public async createFolder(folderRequest: FolderRequest, createdBy: string): Promise<Folder | null> {
        const folder = {
            ...folderRequest,
            createdBy,
        } as Folder;
        return this.folderRepository.createFolder(folder);
    }


    public async updatePatientProfileId(folderId: string, patientProfileId: string): Promise<Folder | null> {
        const patientProfile = await this.patientService.findById(patientProfileId);
        if (!patientProfile) {
            throw new Error('Patient profile not found with id: ' + patientProfileId);
        }
        return this.folderRepository.updatePatientProfileId(folderId, patientProfileId);
    }


    public async getFolderOfUser(userId: string): Promise<FolderResponse[] | null> {
        const folders = await this.folderRepository.findFoldersByCreatedBy(userId) as Folder[];
        if (!folders) {
            return null;
        }
        // map folder to folder response
        const folderResponses: FolderResponse[] = folders.map(folder => ({
            id: folder.id,
            title: folder.title,
            description: folder.description,
            chatSessionIds: folder.chatSessionIds,
            patientProfileId: folder.patientProfileId,
            createdBy: folder.createdBy,
            isDeleted: folder.isDeleted,
            createdDate: folder.createdDate,
            updatedDate: folder.updatedDate,
            chatSessionsInfo: [],
        }));

        for (const folderResponse of folderResponses) {
            for (const chatSessionId of folderResponse.chatSessionIds || []) {
                const chatSession = await this.chatSessionRepository.findById(chatSessionId);
                if (chatSession) {
                    folderResponse.chatSessionsInfo?.push({
                        id: chatSession.id,
                        title: chatSession.title,
                        isDeleted: chatSession.isDeleted,
                        createdDate: chatSession.createdDate,
                        updatedDate: chatSession.updatedDate,
                    });
                }
            }
        }
        return folderResponses;
    }
    

    public async findById(folderId: string): Promise<Folder | null> {
        return this.folderRepository.findById(folderId);
    }

    
}

File: servers/DoctorService/Services/Folder/FolderService.cs
using DoctorService.Models;
using DoctorService.Repositories.Folder;
using DoctorService.Web.Requests;
using DoctorService.Web.Responses;

namespace DoctorService.Services.Folder
{
    public class FolderService : IFolderService
    {
        private readonly IFolderRepository _folderRepository;

        public FolderService(IFolderRepository folderRepository)
        {
            _folderRepository = folderRepository;
        }

        public async Task<Models.Folder?> CreateFolderAsync(FolderRequest folderRequest, string createdBy)
        {
            var folder = new Models.Folder
            {
                Title = folderRequest.Title,
                Description = folderRequest.Description,
                CreatedBy = createdBy,
                Type = folderRequest.Type
            };

            return await _folderRepository.CreateFolderAsync(folder);
        }

        public async Task<Models.Folder?> UpdatePatientProfileIdAsync(string folderId, string patientProfileId)
        {
            // Note: In the Node.js version, there's a patient service validation
            // For now, we'll implement the basic update. You may want to add patient validation later
            return await _folderRepository.UpdatePatientProfileIdAsync(folderId, patientProfileId);
        }

        public async Task<Models.Folder?> FindByIdAsync(string folderId)
        {
            return await _folderRepository.FindByIdAsync(folderId);
        }

        public async Task<List<FolderResponse>?> GetFolderOfUserAsync(string userId)
        {
            var folders = await _folderRepository.FindFoldersByCreatedByAsync(userId);
            if (folders == null)
            {
                return null;
            }

            // Map folders to folder responses
            var folderResponses = folders?.Select(folder => new FolderResponse
            {
                Id = folder.Id,
                Title = folder.Title,
                Description = folder.Description,
                ChatSessionIds = folder.ChatSessionIds,
                PatientProfileId = folder.PatientProfileId,
                CreatedBy = folder.CreatedBy,
                IsDeleted = folder.IsDeleted,
                CreatedDate = folder.CreatedDate,
                UpdatedDate = folder.UpdatedDate,
                ChatSessionsInfo = new List<ChatSessionInfo>() // TODO: Implement chat session info retrieval
            }).ToList() ?? new List<FolderResponse>();

            return folderResponses;
        }

        public async Task<Models.Folder?> UpdateFolderAsync(string folderId, UpdateFolderRequest request)
        {
            return await _folderRepository.UpdateFolderAsync(folderId, request.Title, request.Description);
        }

        public async Task<bool> DeleteFolderAsync(string folderId)
        {
            return await _folderRepository.SoftDeleteFolderAsync(folderId);
        }
    }
}

